default-fields/0000-default-fields.md

```diff
@@ -6,31 +6,268 @@
+
+```rust
+mod data {
+    pub struct Foo {
+        pub a: &'static str,
+        pub c: i32,
+    }
+}
+
+let foo = data::Foo {
+    a: "Hello",
+    c: 42,
+}
+```
+
+We can add a new field `b` to this structure with a default value, and the calling code doesn't need to change:
+
+```rust
+mod data {
+    pub struct Foo {
+        pub a: &'static str,
+        pub b: bool = true,
+        pub c: i32,
+    }
+}
+
+let foo = data::Foo {
+    a: "Hello",
+    c: 42,
+}
+```
+
+By using field defaults, callers can use structure literals without having to know about any private fields:
+
+```rust
+mod data {
+    pub struct Foo {
+        pub a: &'static str,
+        pub c: i32,
+        private_field: bool = true
+    }
+}
+
+let foo = data::Foo {
+    a: "Hello",
+    c: 42,
+}
+```
 
 # Drawbacks
 [drawbacks]: #drawbacks
 
-Why should we *not* do this?
+Field defaults are limited to `const` expressions. This means there are values that can't be used as defaults, such any value that requires allocation, including common collections like `Vec`.
 
 # Alternatives
 [alternatives]: #alternatives
 
-What other designs have been considered? What is the impact of not doing this?
+## Allow arbitrary expressions instead of just constants
 
-# Unresolved questions
-[unresolved]: #unresolved-questions
+Allowing arbitrary expressions as field defaults would make this feature more powerful. However, limiting field defaults to `const`s maintains the expectation that struct literals are cheap and deterministic. The same isn't true when arbitrary expressions that could reasonably panic or block on io are allowed.
+
+For complex initialisation logic, builders are the preferred option because they don't need to carry this same expectation.
 
-What parts of the design are still TBD?
+## Allow `Default::default()` instead of just constants
+
+An alternative to allowing any expression as a default value is allowing `Default::default()`, which is expected to be cheap and deterministic:
+
+```rust
+struct Foo {
+    a: &'static str,
+    b: Vec<i32> = Vec::default(),
+    c: i32,
+}
+
+let foo = Foo {
+    a: "Hello",
+    c: 42,
+}
+```
+
+It could be argued that supporting `default()` is an artificial constraint that doesn't prevent arbitrary expressions. The difference is that `Default` has an expectation of being cheap, so using it to inject logic into field initialisation is an obvious code smell.
+
+Allowing functionality to be injected into data initialisation through `default()` means struct literals may have runtime costs that aren't ever surfaced to the caller. This goes against the expectation that literal expressions have small, predictable runtime cost and are totally deterministic (and trivially predictable).
+
+## Type inference for fields with defaults
+
+Reduce the amount of code needed to define a structure by allowing type inference for field defaults:
+
+```rust
+struct Foo {
+    a: &'static str,
+    b = 42,
```

Comment:
Integers are not the best example for this because of the more complex type inference, maybe better to use a string or a struct.